### PR TITLE
feat: List V2

### LINKED ISSUE #5245: [Bug] Deploy Preview - Widget Error does not get displayed when added into the List widget however, when placed outside the widget the error gets displayed.
## Description
Deploy Preview - It is observed that Widget Error does not get displayed when added into the List widget however, when placed outside the widget the error gets displayed.

**Expected Result:**
Widget error should be displayed inside list widget 

**Actual Result:**
Select widget error is not displayed inside list widget

### Steps to reproduce the behaviour:

1.Launch Appmsith app in the Deploy preview - https://deploy-preview-4189--appsmith-editor-ce.netlify.app/
2.Drag and Drop list widget
3.Place Select widget inside an item container 
4.Open property pane of select widget 
5.Copy the existing select query and paste it again to duplicate and observe no error message is displayed 
6.Drag and Drop select widget outside list widget 
7.Open property pane of select widget 
8.Copy the existing select query and paste it again to duplicate and observe

**VideoLink:**
[LOOM DEMO](https://www.loom.com/share/a1a2ccd19fc14411bebc765bb629fc67)

### Important Details

Version: [14.1 (16611.1.21.161.3)]
OS: [MacOS]
Browser [Safari]
Environment [Deploy Preview]

### LINKED ISSUE #6775: [Bug] List widget: Probable re-rendering due to eval in child
## Description

The list widget started behaving abruptly in a use case of a chart widget as one of the child. Adding necessary details before

https://user-images.githubusercontent.com/11089579/130358602-6100c41d-9a69-461b-9951-44aef98514e3.mov

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [MacOSX]
- Browser [chrome]
- Environment [production]

### LINKED ISSUE #6840: [Bug] Dropdown selectedOptionValue cannot be accessed inside List Widget
## Description

The selected option value of the dropdown cannot be accessed inside the list widget
1. Drag a List
2. Add a dropdown inside it
3. Drag a Text Widget
4. Bind the text widget to {{List1.items[0].Select1.selectedOptionValue}}

### LINKED ISSUE #6858: [Feature] Allow selecting the List Item Container
## Description

Because we cannot select the list item container, we cannot conditionally color the items of the list

### LINKED ISSUE #8237: [Bug] The list widget does not reset to the initial page, neither does the selection clears
## Description

When we try to reset a list widget on a button click and it is observed that the page number and the item selection does not reset, However the child widgets does reset.

### LINKED ISSUE #9568: [Bug]: cannot reach widgets inside a list if the list has a "OnListItemClick" property specified.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When the list has some widgets inside, and something specified in the "OnListItemClick" property. Now you cannot select the widgets inside the list. When you attempt to, the OnListItemClick gets called instead. :(

Uploading Screen Recording 2021-12-04 at 9.27.37 PM.mov…

### Steps To Reproduce

1. I have created an app to reproduce this: https://app.appsmith.com/applications/61ab8e37d3e4214454c23c86/pages/61ab8e37d3e4214454c23c88/edit
2. see there are 2 lists: List1 and List2
3. List1 has OnListItemClick property to open some modal. List2 has no OnListItemClick event specified. 
4. When you try to access some widget inside List1, the modal opens and the property pane also shows the properties of the modal. You cannot anymore access the properties of the widgets inside the list unless you remove the OnListItemClick property and reset it to "no action"
5. You can do all this with List2, I'm guessing because it has no OnListItemClick specified. 

### Environment

Production

### Version

Cloud

### LINKED ISSUE #9607: [Bug]: List widget does not reset the meta data for input widget first list item on data change.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Usecase from user:

`I have a list widget. Each item in that widget has an input field. Data for that list is coming from appsmith store. I have a button on page which when clicking changes data in appsmith store which in turn gives updates to list widgets data. The problem is that if I edit input field in list widget and then click button on page to regenerate that list widgets data, I can still see on first item in the list that input field has the same value as I edited before when list widget had another data which is not correct... I would want to reload completely list widget when I change data for it, how could I do that?`

### Steps To Reproduce

1. Add list widget.
2. Add input widget to the list widget
3. Add a button out of the list widget and add an action on that button that will change the list data.
4. Now enter anything in the input of the first input and click on the button.
5. Notice the input still have the same value even on data change.

### Environment

Production

### Version

Cloud

### LINKED ISSUE #10841: [Bug]: Evaluation failed for rating widget's Tooltips property
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Screenshot 2022-02-02 at 9 48 20 PM

### Steps To Reproduce

Please fork this application (`populate list from query` page)
https://app.appsmith.com/applications/61fa19d411f4c46e3dfd2250/pages/61fa19d411f4c46e3dfd2253
### Environment

Production

### Version

Cloud

### LINKED ISSUE #11332: [Bug]: List widget - When using pagination, no way to go back to previous page after reaching last page
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When using server-side pagination on the list widget, if you click on the Next button after the last page with data, it shows "No data to display" and then there is no way to go back to the previous page and the list widget stays stuck on the no data page.

Uploading Screenrecording_20220222_084207.mp4…

.

### Steps To Reproduce

1. Connect data to a list widget and use server-side pagination
2. Go to the last page with data and click Next
3. It shows a page with the text "No data to display" 
4. Now there is no way to go back to the previous page

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11739: [Bug]: List Widget's PageNo does not reflect correctly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

.png :-
Screenshot 2022-03-09 at 5 17 30 PM

### Steps To Reproduce

1. Create a ListWidget with data spanning multiple pages
2. Map the ListWidget's PageNo to Input text to see the value changing as user moves pages
3. Observe that the values are not computed correctly & remains the same as page 1
4. **Expected**: Values should be computed rightly

### Public Sample App

_No response_

### Version

Cloud Release v1.6.14

### LINKED ISSUE #13211: [Bug]: PropertyPane evaluation of Child widget in a list widget  throws an error when some chars are used
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Text widget in a list widget has 's and that throws an error for some reason

### Steps To Reproduce

Drag a Text widget into a List Widget, Then change the text in the Text widget to `Yesterday's`.
Observe an `SyntaxError: Unexpected identifier` error 

### Public Sample App

_No response_

### Version

cloud

### LINKED ISSUE #14655: [Bug]: Inputs inside Lists automatically add values on different pages in the Run modal of the Saas Manager App.
On the Run modal (SaaS Manager App), when you write a data in the first input and move to the next page, that data is already in the first input. Only the first input is overwritten, the others work fine. 

The following video shows the problem:

https://images.zenhubusercontent.com/275993885/a10e18f8-c30c-4b4f-b4a1-834cbc416dde/video1994857732.mp4

### LINKED ISSUE #14697: [Bug]: Update on Input widget inside List shows that value on all pages of the corresponding row until browser refresh
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

On the List widget, if there is an Input widget, note the following the bugs

1. When you first bind the data to the list widget, the value in the first input widget will be blank until the browser is refreshed.
Screenshot 2022-06-21 at 9 55 27 AM

2. If you update the first input widget on the list, the display shows that updated value in all pages of the corresponding row until the page is refreshed. In the screenshot below, I change the name Jimmy on page 1 and the display reflects on all the other pages until the browser is refreshed. When serverside pagination is not used, this behaviour only occurs with the first item on the list 
Screenshot 2022-06-21 at 10 04 08 AM

3. When using serverside pagination on list, the bug in step 2 occurs for every list item on the page, not just the first one
Screenshot 2022-06-21 at 10 08 54 AM

I only checked this with update query, this probably also happens on an insert query

### Steps To Reproduce

1. Add a list widget to the canvas
2. Add an input widget
3. Bind data to the list and display a field on the input widget. You will notice the input on the first list item is blank until the page is refreshed
4. Change the value on the input widget on the first list item and run an update query
5. Navigate to the other pages in the list and note that it shows the same updated value on all the pages until the browser refresh
6. Check the same with serverside pagination on, you will note this behaviour on all the list items and not just on first list item on the current page

### Public Sample App

https://app.appsmith.com/app/input-widget-bugs-inside-list/page1-62b14dc740b36d162478b0da

### Version

v1.7.4

### LINKED ISSUE #15094: [Bug]: Dragging a List Widget Inside Another List Widget Crashes the App
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Although currently, nested lists are not supported, if the user resizes a list, they can make it fit inside another list, which makes the app crash.
Users should not be allowed to add a list widget inside another list widget unless [nested lists](https://github.com/appsmithorg/appsmith/issues/14341) will be supported.

Loom screen-recording: https://www.loom.com/share/88ff801b01154010996fb73dcb261f90

https://app.intercom.com/a/apps/y10e7138/inbox/inbox/conversation/164629100139777#part_id=initial-part-164629100139777-1365134472

### Steps To Reproduce

1. Add a List widget to the canvas.
2. Add a second List widget to the canvas. 
3. Resize one of the List widgets, decreasing its size until it fits inside the item of the other List widget.
4. Drag it inside the "parent" List widget and drop it once it's not highlighted with red anymore.
5. Note that the app crashes and console errors get displayed.

### Public Sample App

_No response_

### Version

Cloud v1.7.7-SNAPSHOT

### LINKED ISSUE #15487: [Bug]: Only the first list item resets with the resetWidget function
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

A user is performing bulk uploads using the list widget; after he submits the form, he wants to reset his input widgets on the list items. He used the `resetWidget` function and passed the `List1` widget with `children: true`, but it does not work.

Link: https://community.appsmith.com/t/resetting-a-list-in-a-modal/1230/2

### Steps To Reproduce

1. Drag and drop a list widget
2. Add a couple of input widgets to the first list item
3. Add a button and set the `onClick` to `resetWiget` - `{{resetWidget("List1",true)}}`
4. Write some dummy text on the input widgets
5. Click the button to reset the list
6. Only the data on the first list items resets

### Public Sample App

_No response_

### Version

Cloud

Front logo [Front conversations](https://app.frontapp.com/open/top_5zax8)

### LINKED ISSUE #15988: [Bug]: On changing Box radius/shadow of widgets added inside list vanishes the data on copy pasting list widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The data vanished on copy pasting list widget whenever Box radius/shadow is being changed on widgets added inside List widget.

Similar issue [#17815](https://github.com/appsmithorg/appsmith/issues/17815).

https://www.loom.com/share/f6f68a11cd4442aa81659c32f3056960

### Steps To Reproduce

1. Drag & Drop List widget
2. Click on image widget which is already added inside List widget
3. Change Box radius of image widget from property pane
4. Now copy the list widget and paste it
5. Observe that data populated on List widget is vanished from both List widget

### Public Sample App

_No response_

### Version

Cloud / Production

### LINKED ISSUE #16576: Task: List v2 - Child widget property pane, {{currentItem}} binding evaluated value is always undefined
- [ ] It should show the evaluated value instead of `undefined`
- [ ] It should throw validation errors.

### LINKED ISSUE #16582: Task: List v2: Implement hashed key
In list v2 every row has to be identified by a unique key, the key can by any unique value on the list data.

If no key is selected then the whole list data item has to be hashed using non-cryprographic algorithm (for fastest hash generation) to generate unique keys for each row.

A warning also should be added to the property depicting the importance of the property.

### LINKED ISSUE #16586: Task: List v2 - Page size derived property fix
The list v2 on mount shows erratic number for the page size property. The expectation here is that if all the relevant information is not present while generating `pageSize`, its best to make it undefined rather than to give an estimated number as it will be used to generate meta widgets and also used in server side rendering

### LINKED ISSUE #16788: Task: List v2 - Jest tests for MetaWidgetGenerator

### LINKED ISSUE #16898: [Bug]: Code mirror hint suggesting wrong types in List widget Property pane
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

I have a list widget which is populated with an `array` of `JSON Objects`. Each `JSON Object` has a sub-array of `JSON objects` called "Tags".
The Autocomplete identifies `Tags` as a string instead of `Array`

### Steps To Reproduce

aasmith-array
appsmith_array_2

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #17386: Task: List v2 - Access of currentRow/currentItem/levels in all executeAction calls

### LINKED ISSUE #17427: [Bug]: List widget | onPageSizeChange action gets triggered without changing the page size upon editing the app
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Upon editing an app , the action set for the `onPageSizeChange` property gets triggered even when we don't change the page size.

https://www.loom.com/share/14b0ee40de23415fa893a06eab20023c

### Steps To Reproduce

1. D&D a List widget.
2. Configure an action for the `onPageSizeChange` property.
3. Now go to the homepage and visit the app again.

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #17480: Task: List v2 - Enable ListWidgetV2 APIs.
Add the following APIs to the List widget to help users access data from the List widget.

```
listData
  Singleton - listDataItem

item -> from listData
row  -> from the list row i.e widgets

selectedItemIndex       -> selectedRowIndex
selectedItem            -> selectedItem 
selectedViewItemIndex   -> selectedRowViewIndex (Internal)
currentSelectedViewItem -> selectedRow
currentViewItems        -> currentViewRows
                        -> triggeredRowIndex
                        -> triggeredRowViewIndex (Internal)
                        -> triggeredRow
onListItemClick         -> onRowClick
currentItem             -> currentItem
currentIndex            -> currentIndex
currentRow              -> currentRow

level (TBD)

```

**NB:** The following should work for Pagination and Infinite Scrolling

Front logo [Front conversations](https://app.frontapp.com/open/top_5zhjg)

### LINKED ISSUE #17555: Task: List v2 - Autocomplete suggestion
- Nested List levels - level_1, level_2 etc
- currentItem
- currentRow
- currentIndex
- currentViewItems

### LINKED ISSUE #17777: Task: List v2 - Rename property gridGap to itemSpacing.

### LINKED ISSUE #17901: Task: List v2 - Cypress tests
Add new tests and migrate List V1 tests to V2

### LINKED ISSUE #18105: Task: List v2 - Server side pagination - flashing rows on page change
When server side pagination is enabled, and the next page is clicked; the rows flash multiple times before settling on the new page rows
https://www.loom.com/share/837e0f2ed750498ea89464d560d7b53b

### LINKED ISSUE #18106: Task: List v2 - Canvas meta widget is generated multiple times
If a debugger is placed just before `modifyMetaWidgets` call, it is found that the canvas widget with the same props is present in `addOrUpdate` option of `modifyMetaWidgets`. This will cause to execute unnecessary evaluations.

### LINKED ISSUE #18110: [Bug]: Error 'currentItem is not defined' When Running Multiple Queries with currentItem Sent as Parameter onListItemClick in a List Widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When running multiple queries in the onListItemClick event of a List widget and sending currentItem as parameter, only the first query runs correctly. The next one fails, showing the error `Uncaught Promise Rejection: currentItem is not defined`.
image
The workaround for this issue is to create a JS Function inside a JS Object and move the code there, then run this JS function in the onListItemClick event. The user can send `currentItem` to the function as an argument to be able to access its properties.

### Steps To Reproduce

1. Add a List widget to the canvas.
2. Run multiple queries in the onListItemClick event.
3. Note that only the first query gets executed correctly and the error shows  up.

### Public Sample App

https://app.appsmith.com/app/list-currentitem/page1-636568134c80470350f619ea

### Issue video log

_No response_

### Version

Appsmith v1.8.7-SNAPSHOT

Front logo [Front conversations](https://app.frontapp.com/open/top_5nmks)

### LINKED ISSUE #18206: [Bug]: Issue with evaluation of pageSize with list v1 on reload
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

https://www.loom.com/share/bf95456e9b064835895eaede04daa4cf

### Steps To Reproduce

1. Dnd a List widget 
2. Dnd a text widget and bind it with List page size to check the exact page size.
3. Move to another page say Postgres Query and add binding {{List1.pageSize}}
4. refresh the page and check the results in the query (You will see pageSize is coming as 2 instead of 3) but in the main canvas the pageSize is coming as 3 only
For more details check LOOM

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

Production

### LINKED ISSUE #18287: Task:  Change List widget's APIs
`gridGap`      --> `rowGap`
`rowIndex`  ---> `viewIndex` in MetaWidgetGenerator
`index` ---> `rowIndex` in MetaWidgetGenerator

```
selectedItemIndex       -> selectedRowIndex       -> selectedItemIndex
selectedItem            -> selectedItem           -> selectedItem
currentSelectedViewItem -> selectedRow            -> selectedItemView
currentViewItems        -> currentViewRows        -> currentItemsView
                        -> triggeredRowIndex      -> triggeredItemIndex
                        -> triggeredRow           -> triggeredItemView

```

### LINKED ISSUE #18340: Task: Misc. fixes 1
- [x] View mode rows have opacity
- [x] Vertical scrollbars appear on simple lists
~Container size keeps decreasing when infinite scroll is toggled~ TBD
- [x] New widget added to list does not work in nested list level_1.currentRow 
~Explore warning - Widget property pane config not defined CANVAS_WIDGET~ Not introduced by List v2 
- [x] Update new property pane sections

### LINKED ISSUE #18382: Task: Remove Meta widgets from autocomplete and debugger
Remove all meta widgets from Autocomplete

Screenshot 2022-11-22 at 10 45 46

### LINKED ISSUE #18523: [Bug]: Pagination control issue in List v2
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Issue 1: The background color of the pagination seems to be filled with blue color instead of keeping it white
Issue 2: On hover to the pagination, a tooltip appears which is weird in style. Also if we can remove this tooltip as it has no use as a user

### Steps To Reproduce

1. Drag & Drop List v2 on canvas
2. Click on pagination
3. Observe there is a filled blue color for selected page( Blue color can be observed on any theme set)
4. Hover on pagination
5. Observe a tooltip design appears on hover

### Public Sample App

_No response_

### Issue video log

https://bthrujcsw8.vmaker.com/record/W24M7Pat4SfYeIuu

### Version

Cloud

### LINKED ISSUE #18621: Task: List v2 - First row not editable for page > 1 when server side rendered.
For a server side enabled List widget, the first item of each page is editable i.e the property pane can be accessed to edit the properties of the widgets.
This is not true for client side enabled where List where only the page 1 first item is enabled for editing.

### LINKED ISSUE #18638: [Bug]: Resetting ListV2 data causes page number to be 0
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The Page number resets and changes to Zero which displays no widget in the list widget

### Steps To Reproduce

https://user-images.githubusercontent.com/31691737/205287546-08309323-b8f3-4493-a37d-83cc31d89c86.mov

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

cloud

### LINKED ISSUE #18730: [BUG]: List widget V2 SelectedItemView and TriggeredItemView in Serverside Pagination

### LINKED ISSUE #19043: Task: List v2 Prevent duplicated template widget id generation for meta widgets

### LINKED ISSUE #19150: Task: List v2 Performance profiling for evaluation

### LINKED ISSUE #19437: [Bug]: Form widget crashes in List widget v2

### LINKED ISSUE #19498: [BUG] Listv2 canvas widget not accepting child when there's space

### LINKED ISSUE #19503: Task: List v2 disable auto height property in all the widgets of the List widget

### LINKED ISSUE #19648: [Bug]: Misc. bugs 2
1. Page number hover shows blue color on text - https://www.loom.com/share/33a0e78f9b3a45e7a51f0cc3ea23f9cf
2. Unwanted flickering - https://www.loom.com/share/3470faa7136b40999ce9707ea40bfac3
3. ~Label z index~ - see image. Update: This is reproducible with normal container so categorising this as non List widget bug. 

4. All the list containers on hover become transparent - https://www.loom.com/share/01c81297e75d4cc08849870d38cde743
5. ~Keyboard tabbing doesn't work between items in a list - Like if you reach end of item 1, it doesn't go to item 2~
  Update: This is happens when an image widget comes in between 2 inputs. This can be reproduced outside List widget so categorising this as non List widget bug.

### LINKED ISSUE #19652: [BUG]: Incorrect currentItem value due to evaluationSubstitutionType

### LINKED ISSUE #19739: [Task]: Refactor List Widget V2 Api
1. Add `triggereitem` to List v2
2. Rename `onRowClick` to `onItemClick` on List v2

### LINKED ISSUE #19808: [Bug]: Unresponsive page when List widget's container is copy pasted
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When list widget is dropped on the canvas and container of list widget is copied a message is shown "This widget cannot be copied" however when you paste it the widgets gets overlapped with the existing widgets. When user pastes it the second time the page breaks.
https://loom.com/share/2ac9de56bcbe4aeb8896c50bdde42897

Scenario 2:
Drop a Tab widget now drop a List widget inside tab1. Now in the entity explorer expand the tab widget and notice Canvas3, Container & Canvas4 widgets appear. Click on Canvas 5 the property pane opens up > Copy and paste it on the canvas - the page becomes unresponsive after this action. Whereas when Canvas4 is copy pasted there is nothing seen on the canvas but in the entity explorer Canvas4Copy is seen.
(https://loom.com/share/a20f146436a54227bef166c5d1ffdc8f)

As seen on Windows machine

### Steps To Reproduce

1. Drag and drop List widget
2. In entity explorer click on list widget drop down 
3. Click on Container1 > Copy > Observe widget gets overlapped > Paste again 

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

cloud

### PR DESCRIPTION
**Pull request by ashit-rath**
## Description

TL;DR
This is a complete architectural change of of List widget works to support all widgets we currently have and should automatically support any future widgets.
It also introduces nested List widgets i.e a list widget can have a another list widget which in turn can have another list widget.

Fixes #18206
Fixes #6775
Fixes #13211
Fixes #16582
Fixes #11739
Fixes #15094
Fixes #6840
Fixes #10841
Fixes #17386
Fixes #18340
Fixes #16898
Fixes #17555
Fixes #6858
Fixes #9568
Fixes #17480
Fixes #18523
Fixes #18206  
Fixes #16586
Fixes #18106
Fixes #16576
Fixes #14697
Fixes #9607
Fixes #19648 
Fixes #19739
Fixes #19652 
Fixes #18730 
Fixes #19503 
Fixes #19498
Fixes #19437
Fixes #5245 
Fixes #19150
Fixes #18638
Fixes #11332
Fixes #17901
Fixes #19043
Fixes #17777
Fixes #8237
Fixes #15487
Fixes #15988
Fixes #18621
Fixes #16788
Fixes #18110
Fixes #18382
Fixes #17427
Fixes #18105
Fixes #18287
Fixes #19808
Fixes #14655

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Cypress
- Jest
- Manual

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ashit-rath** [comment]
@arunvjn We need you help in reviewing certain parts we made changes to. Feel free to explore other bits that might not be mentioned here and provide callouts.
```
app/client/src/ce/workers/Evaluation/evaluationUtils.ts
app/client/src/sagas/EvaluationsSaga.ts
app/client/src/sagas/PostEvaluationSagas.ts
app/client/src/selectors/debuggerSelectors.tsx
```

File: app/client/src/sagas/EvaluationsSaga.ts [hunk 1]
```diff
@@ -165,6 +167,7 @@ export function* evaluateTreeSaga(
     allActionValidationConfig,
     requiresLinting: isEditMode && requiresLinting,
     forceEvaluation,
+    metaWidgets,
```
Comment (arunvjn) [inline app/client/src/sagas/EvaluationsSaga.ts:170]: I don't understand how `metaWidgets` is being used by the worker. I don't see references to it in the handler. Could you please check ?

File: app/client/src/sagas/PostEvaluationSagas.ts [hunk 2]
```diff
@@ -359,11 +359,11 @@ export function* updateTernDefinitions(
 
   if (!shouldUpdate) return;
   const start = performance.now();
-  // remove private widgets from dataTree used for autocompletion
-  const treeWithoutPrivateWidgets = getDataTreeWithoutPrivateWidgets(dataTree);
+  // remove private and suppressAutoComplete widgets from dataTree used for autocompletion
+  const filteredDataTree = getFilteredDataTree(dataTree);
```
Comment (arunvjn) [inline app/client/src/sagas/PostEvaluationSagas.ts:363]: Would recommend a more specific function name here if it is not used elsewhere, something like `getDataTreeForAutocomplete`, and remove the comment.

File: app/client/src/sagas/EvaluationsSaga.ts (hunk 1 above)
Comment (ashit-rath) [reply app/client/src/sagas/EvaluationsSaga.ts:170]: I don't think it is used in the handler. Are the canvas widgets in the worker used only for replay logic or something else as well?

File: app/client/src/sagas/PostEvaluationSagas.ts (hunk 2 above)
Comment (ashit-rath) [reply app/client/src/sagas/PostEvaluationSagas.ts:363]: Agreed!

File: app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx [hunk 3]
```diff
@@ -349,7 +350,7 @@ const PropertyControl = memo((props: Props) => {
     // is changed on propertyPane. For e.g - set/update parent property
     if (childWidgetPropertyUpdateEnhancementFn) {
       const hookPropertiesUpdates = childWidgetPropertyUpdateEnhancementFn(
-        widgetProperties.widgetName,
+        widgetProperties,
```
Comment (ashit-rath) [inline app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx:353]: I think we should revert this change. This was done to support v2 but eventually dropped this enhancement use thus the change might not be required
cc: @Tooluloope

**Comment by ashit-rath** [comment]
@SatishGandham 
```
app/client/src/components/editorComponents/EditorContextProvider.tsx
app/client/src/components/editorComponents/EditorContextProvider.test.tsx
app/client/src/constants/WidgetConstants.tsx
app/client/src/entities/DataTree/dataTreeFactory.ts
app/client/src/pages/AppViewer/index.tsx
app/client/src/pages/Editor/WidgetsEditor/index.tsx
app/client/src/reducers/entityReducers/canvasWidgetsStructureReducer.ts
app/client/src/sagas/WidgetOperationUtils.ts
app/client/src/utils/canvasStructureHelpers.ts
app/client/src/utils/widgetRenderUtils.tsx
app/client/src/widgets/BaseWidget.tsx
app/client/src/widgets/ContainerWidget/component/index.tsx
app/client/src/widgets/MetaHOC.tsx
// This is used inside List widget to overrider few methods
app/client/src/widgets/MetaWidgetContextProvider.tsx
app/client/src/selectors/editorSelectors.tsx
app/client/src/selectors/debuggerSelectors.tsx
app/client/src/sagas/selectors.tsx
```

**Comment by ashit-rath** [comment]
@ohansFavour Could you review [this](https://github.com/appsmithorg/appsmith/pull/15839/commits/676b4af985c3692527514a6e6aa1e356f57935ff)

File: app/client/package.json [hunk 4]
```diff
@@ -22,6 +22,7 @@
     "@manaflair/redux-batch": "^1.0.0",
     "@sentry/react": "^6.2.4",
     "@sentry/tracing": "^6.2.4",
+    "@tanstack/virtual-core": "^3.0.0-beta.18",
```
Comment (SatishGandham) [inline app/client/package.json:25]: We are already using react window, for few widgets. Are these two related, can we use one library.

Comment (ashit-rath) [reply app/client/package.json:25]: We needed a headless virtualizer as we were dealing with direct widgets and required heavy customisation; on the other hand react-window had it's own component which controlled height widget and a bunch of other things which was quite restrictive and might cause problems in future. So we opted for a headless one.

File: app/client/src/reducers/entityReducers/metaWidgetsReducer.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+const initialState: MetaWidgetsReduxState = {};
+
+const metaWidgetsReducer = createImmerReducer(initialState, {
+  [ReduxActionTypes.MODIFY_META_WIDGETS]: (
+    state: MetaWidgetsReduxState,
+    action: ReduxAction<ModifyMetaWidgetPayload>,
+  ) => {
+    const {
+      addOrUpdate,
```
Comment (SatishGandham) [inline app/client/src/reducers/entityReducers/metaWidgetsReducer.ts:54]: Please add comments to what these properties are, its hard to follow without context.
May be in types file.

File: app/client/src/widgets/withWidgetProps.tsx [hunk 6]
```diff
@@ -33,21 +36,39 @@ function withWidgetProps(WrappedWidget: typeof BaseWidget) {
...
+      hasMetaWidgets,
+      referencedWidgetId,
+      requiresFlatWidgetChildren,
+      skipWidgetPropsHydration,
+      type,
+      widgetId,
+    } = props;
     const isPreviewMode = useSelector(previewModeSelector);
 
     const canvasWidget = useSelector((state: AppState) =>
```
Comment (SatishGandham) [inline app/client/src/widgets/withWidgetProps.tsx:50]: Widget could be either canvasWidget or meta widget, I think this will be simpler

```
    const widget = useSelector((state: AppState) =>
      getWidget(state, widgetId),
    ) ||  useSelector(getMetaWidget(widgetId));
```

or 

```
    let widget = useSelector((state: AppState) =>
      getWidget(state, widgetId),
    );
    if(!iwdget){widget = useSelector(getMetaWidget(widgetId))}
```

File: app/client/src/widgets/withWidgetProps.tsx [hunk 7]
```diff
@@ -33,21 +36,39 @@ function withWidgetProps(WrappedWidget: typeof BaseWidget) {
...
       getWidget(state, widgetId),
     );
+    const metaWidget = useSelector(getMetaWidget(widgetId));
```
Comment (SatishGandham) [inline app/client/src/widgets/withWidgetProps.tsx:53]: Widget could be either canvasWidget or meta widget, I think this will be simpler

```
    const widget = useSelector((state: AppState) =>
      getWidget(state, widgetId),
    ) ||  useSelector(getMetaWidget(widgetId));
```

or 

```
    let widget = useSelector((state: AppState) =>
      getWidget(state, widgetId),
    );
    if(!iwdget){widget = useSelector(getMetaWidget(widgetId))}
```

Comment (SatishGandham) [reply app/client/src/widgets/withWidgetProps.tsx:53]: You are already doing it at line 91, lets move it a little up.

File: app/client/package.json (hunk 4 above)
Comment (SatishGandham) [reply app/client/package.json:25]: Does this lib have any perf benefits over react window, should we replace react window with this when we get a chance?

File: app/client/src/widgets/TableWidgetV2/component/header/actions/filter/FilterPane.tsx [hunk 8]
```diff
@@ -119,10 +120,14 @@ class TableFilterPane extends Component<Props> {
 }
 
 const mapStateToProps = (state: AppState, ownProps: TableFilterPaneProps) => {
+  const widgetLikeProps = {
```
Comment (SatishGandham) [inline app/client/src/widgets/TableWidgetV2/component/header/actions/filter/FilterPane.tsx:123]: What is widget like props?

File: app/client/package.json (hunk 4 above)
Comment (ashit-rath) [reply app/client/package.json:25]: I didn't find much evidence of performance between the two and they both boast of lightning perf. 
https://johnnyreilly.com/2020/05/10/from-react-window-to-react-virtual

File: app/client/src/widgets/TableWidgetV2/component/header/actions/filter/FilterPane.tsx (hunk 8 above)
Comment (ashit-rath) [reply app/client/src/widgets/TableWidgetV2/component/header/actions/filter/FilterPane.tsx:123]: As `getWidgetMetaProps` now needs `widgetProps` instead of just `widgetId`, we created an object `widgetLikeProps` which contains just `widgetId` as that is the base min props required and then coerce it as `WidgetProps`. Hence the name `widgetLikeProps`

File: app/client/src/widgets/withWidgetProps.tsx (hunk 7 above)
Comment (ashit-rath) [reply app/client/src/widgets/withWidgetProps.tsx:53]: We cannot use hooks in conditions so this might not be possible. 
Apologies for the late reply!

**Review by SatishGandham — APPROVED** [review]

**Review by Tooluloope — APPROVED** [review]

**Review by Tooluloope — APPROVED** [review]
